src/sentry/taskworker/scheduler.py

```diff
@@ -0,0 +1,255 @@
+
+        If a key is already set this method will raise an error.
+        """
+
+    @abc.abstractmethod
+    def read(self, taskname: str) -> datetime | None:
+        """
+        Retrieve the last run time of a task
+        Returns None if last run time has expired or is unknown.
+        """
+
+    @abc.abstractmethod
+    def delete(self, taskname: str) -> None:
+        """remove a task key - mostly for testing."""
+
+
+class RedisRunStorage(RunStorage):
+    def __init__(self, redis: RedisCluster[bytes] | StrictRedis[bytes]) -> None:
+        self._redis = redis
+
+    def _make_key(self, taskname: str) -> str:
+        return f"tw:scheduler:{taskname}"
+
+    def set(self, taskname: str, next_runtime: datetime) -> bool:
+        now = timezone.now()
+        duration = next_runtime - now
+
+        result = self._redis.set(self._make_key(taskname), now.isoformat(), ex=duration, nx=True)
+        return bool(result)
+
+    def read(self, taskname: str) -> datetime | None:
+        result = self._redis.get(self._make_key(taskname))
+        if result:
+            return datetime.fromisoformat(result.decode())
+        return None
+
+    def delete(self, taskname: str) -> None:
+        self._redis.delete(self._make_key(taskname))
+
+
+class Schedule(metaclass=abc.ABCMeta):
+    """Interface for scheduling tasks to run at specific times."""
+
+    @abc.abstractmethod
+    def is_due(self, last_run: datetime | None = None) -> bool:
+        """
+        Check if the schedule is due to run again based on last_run.
+        """
+
+    @abc.abstractmethod
+    def remaining_delta(self, last_run: datetime | None = None) -> timedelta:
+        """
+        Get the remaining timedelta until the schedule should run again.
+        """
+
+    @abc.abstractmethod
+    def runtime_after(self, start: datetime) -> datetime:
+        """
+        Get the next scheduled time after `start`
+        """
+
+
+class CrontabSchedule(Schedule):
+    def __init__(self, crontab: crontab) -> None:
+        self._crontab = crontab
+
+    def is_due(self, last_run: datetime | None = None) -> bool:
+        return False
+
+    def remaining_delta(self, last_run: datetime | None = None) -> timedelta:
+        return timedelta(seconds=0)
+
+    def runtime_after(self, start: datetime) -> datetime:
+        return timezone.now()
+
+
+class TimedeltaSchedule(Schedule):
+    def __init__(self, delta: timedelta) -> None:
+        self._delta = delta
+        if delta.microseconds:
+            raise ValueError("microseconds are not supported")
+        if delta.total_seconds() < 0:
+            raise ValueError("interval must be at least one second")
+
+    def is_due(self, last_run: datetime | None = None) -> bool:
+        """Check if the schedule is due to run again based on last_run."""
+        if last_run is None:
+            return True
+        remaining_delta = self.remaining_delta(last_run)
```

Comment:
This seems overly complicated?
```
return now() > (last_run + self._delta)
```

Reply:
We had a simpler solution initially, but ran into issues with microseconds causing tiny diffs in tests. We don't need sub-second precision which is where all the int() rounding comes from.  We could remove the timedelta and return int though to remove some complexity 🤔